Sargon boot

micbakos-rdx opened the pull request:
## Description
This PR uses Sargon OS. Boots with it and reads from/writes to it through its drivers. The first methods from sargon used were
* for new wallet
* for restoring wallet from backup
* for deriving wallet from bdfs and accounts
* for deleting wallet
---

- Encryption helper was deleted since all related methods used for encryption are now living into sargon android. 
- Host Info datastore is migrated from simple shared preferences
- Automatic biometrics are requested when the sargon methods above request encryption or decryption of mnemonics.
- Keystore integrity hopefully fixes the problems we encounter with keystore keyspecs on older version when the user resets the device pin. For more info take a look at https://github.com/radixdlt/sargon/pull/212. This PR will be updated when the sargon PR is merged. (Currently uses a custom build version)

## PR submission checklist
- [X] I have tested account to account transfer flow and have confirmed that it works

giannis-rdx reviewed (commented):
haven't finished the review yet.

I am still reviewing/testing it. 

but the comment I left I think it's important 👇🏽

micbakos-rdx commented on app/src/main/AndroidManifest.xml:
```diff
@@ -83,7 +83,10 @@
             android:value="false" />
 
-
+        <provider
+            android:name="androidx.startup.InitializationProvider"
+            android:authorities="${applicationId}.androidx-startup"
+            tools:node="remove" />
```
Yes, I put this there way before the fix, so I could also test with backups. I will copy the main manifest.

sergiupuhalschi-rdx commented on app/src/main/java/com/babylon/wallet/android/presentation/main/MainViewModel.kt:
```diff
@@ -93,8 +95,8 @@ class MainViewModel @Inject constructor(
-                is ProfileState.Restored -> p2PLinksRepository.observeP2PLinks()
-                else -> p2PLinksRepository.observeP2PLinks().drop(1)
+                is ProfileState.Loaded -> p2PLinksRepository.observeP2PLinks()
+                else -> p2PLinksRepository.observeP2PLinks().drop(1) // TODO why that?
```
This fixed the issue with the connection to the ledger when creating the first account (when the profile doesn't exist yet). If I remember correctly what happened was that during onboarding, when connecting to the ledger we save the p2p link, connect to it from there and it was "conflicting" with the connection logic here and breaking the flow (the connection to ledger wasn't working anymore). I've tried to encompass this in the comment above.

sergiupuhalschi-rdx replied:
I see

sergiupuhalschi-rdx reviewed (approved):
Nice work! Looks very good. I left just a few minor observations.


Tested the scenario I described in the comment regarding p2pLinks (adding a new link from add ledger flow when creating first account) and it works fine.

Also, this might not be related at all with the current PR but since it involved biometric and mnemonic I'm writing it here: when attempting to create an account with ledger, but going back at the point where it requires an actual ledger connection and creating a new wallet without the ledger, "Something went wrong" toast is displayed and nothing happens. I don't see anything in the logs, the error must be swallod somewhere in Sargon. Then if I force close the app and open it again I'm on the home screen with no accounts created (I assume the mnemonic has been created but it failed to create the account?).

Other than this I did a smoke test through the app and it works fine.

giannis-rdx commented on profile/src/main/java/rdx/works/profile/data/repository/ProfileRepository.kt:
```diff
@@ -54,113 +50,146 @@ suspend fun ProfileRepository.updateProfile(updateAction: suspend (Profile) -> P
+
         val profileToSave = profile.copy(
             header = profile.header.copy(
-                lastModified = TimestampGenerator(),
```
I assume you removed this because it is generated in Sargon?

micbakos-rdx replied:
Exactly, on each `set_profile` sargon modifies the timestamp. No need for the host to do the same.

micbakos-rdx commented on app/src/main/java/com/babylon/wallet/android/presentation/settings/troubleshooting/accountrecoveryscan/scan/AccountRecoveryScanViewModel.kt:
```diff
@@ -77,7 +78,7 @@ class AccountRecoveryScanViewModel @Inject constructor(
                 givenTempMnemonic = accessFactorSourcesProxy.getTempMnemonicWithPassphrase()
                 givenTempMnemonic?.let { mnemonic ->
-                    val hostInfo = hostInfoRepository.getHostInfo()
+                    val hostInfo = hostInfoRepository.getHostInfo().getOrThrow()
```
You are right, will do that

micbakos-rdx commented on app/src/main/java/com/babylon/wallet/android/presentation/transaction/composables/ConnectedDAppsContent.kt:
```diff
@@ -44,9 +44,9 @@ import rdx.works.core.domain.DApp
 @Composable
 fun ConnectedDAppsContent(
-    connectedDApps: ImmutableList<Pair<ComponentAddress, DApp?>>,
+    connectedDApps: ImmutableList<Pair<ManifestEncounteredComponentAddress, DApp?>>,
```
No worries @sergiupuhalschi-rdx will do.

micbakos-rdx commented on profile/src/main/java/rdx/works/profile/data/repository/HostInfoRepository.kt:
```diff
@@ -1,137 +1,30 @@
+        val sargonOs = sargonOsManager.sargonOs
 
-        return hostId ?: generateHostId(context)
+        return runCatching { sargonOs.resolveHostId() }
```
@sergiupuhalschi-rdx HostInfo should not fail, and you are right there is no need to propagate the error to the user. On the other hand HostId can only fail if sargon fails to save the newly generated ID (take a look [here](https://github.com/radixdlt/sargon/blob/752616f8bf26fdc071c78c835f5e3976ddc465aa/crates/sargon/src/system/sargon_os/sargon_os.rs#L233)). For this in order to throw this [kotlin driver](https://github.com/radixdlt/sargon/blob/752616f8bf26fdc071c78c835f5e3976ddc465aa/jvm/sargon-android/src/main/java/com/radixdlt/sargon/os/storage/key/HostIdKeyMapping.kt#L38) needs to fail. But in general it shouldn't. What is your opinion on this?

sergiupuhalschi-rdx replied:
By looking at the code, all the reasons it would fail would fall into developer error category, explored these: `newHostIdFromJsonBytes`, `Json.encodeToString`, `deviceStorage.write(..)`. It looks like it's fine to treat it the same as the `HostInfo`. But you definitely have more context and know better.

micbakos-rdx replied:
You are right I have reverted all of such functions to "not throw"

micbakos-rdx commented on profile/src/main/java/rdx/works/profile/domain/backup/RestoreProfileFromBackupUseCase.kt:
```diff
@@ -1,83 +1,37 @@
-                backupProfileRepository.discardTemporaryRestoringSnapshot(backupType)
-                Result.success(Unit)
+        return runCatching {
+            sargonOs.importWallet(
```
These specific functions that mutate the profile, I am not sure are safe to be called in a different thread. The reason for that is how profile is stored [here](https://github.com/radixdlt/sargon/blob/752616f8bf26fdc071c78c835f5e3976ddc465aa/crates/sargon/src/system/sargon_os/profile_state_holder.rs#L24). This is an [`RWLock`](https://doc.rust-lang.org/stable/std/sync/struct.RwLock.html) which I am not sure it is thread safe.

micbakos-rdx replied:
Unfortunately this lock is an `RWLock`, and the way sargon is developed, if such lock blocks the thread it panics and it does not "suspend". 

@GhenadieVP is aware of this, and this will be one of the improvements for `ProfileStateHolder` in the future.

micbakos-rdx commented:
@giannis-rdx @sergiupuhalschi-rdx I have fixed some issues including:

* Updated `set_profile` in sargon to be able to either update or set a new profile.
* Fixed the issue of:
    1. Opening the app from scratch
    2. Opting to create a new wallet
    3. Choosing ledger for the first account
    4. Killing the app in the meantime
    5. Opening the app again, does not navigate to home screen and if you choose to create a new wallet again it creates one successfully.
* Fixed another issue of:
    ... Do the same steps as above until 4
    5. Reset device pattern/pin to a new one
    6. Go back to wallet and create a new one. This should not fail at this time.

giannis-rdx commented on profile/src/main/java/rdx/works/profile/domain/GetProfileUseCase.kt:
```diff
@@ -28,20 +30,15 @@ class GetProfileUseCase @Inject constructor(
-    private fun isInitialized(state: ProfileState): Boolean {
-        return state is ProfileState.Restored && state.hasNetworks()
-    }
+    suspend fun finishedOnboardingProfile(): Profile? = profileRepository.profileState
```
why not `initializedProfile()` or `getInitializedProfile()`? I find it kinda misleading 😄

giannis-rdx replied:
the following is just thoughts - actions not really needed.
I know we already discussed this...

Now, we actually have two functions to get the `Profile` and this is a bit confusing(?). 

I do understand the difference, and I am thinking what if we decide to keep only one (the nullable). But that would cause a lot of nullable checks...

And then what if we keep the `invoke()` and add this piece of code (finishedOnboardingProfile) inside the invoke and throw an exception. Then in the places we must have the initialized profile we catch the error.

giannis-rdx commented on profile/src/main/java/rdx/works/profile/domain/GetProfileUseCase.kt:
```diff
@@ -28,20 +30,15 @@ class GetProfileUseCase @Inject constructor(
-        return state is ProfileState.Restored && state.hasNetworks()
-    }
+    suspend fun finishedOnboardingProfile(): Profile? = profileRepository.profileState
+        .map { state ->
+            (state as? ProfileState.Loaded)?.v1?.takeIf { it.hasNetworks }
+        }
+        .firstOrNull()
```
maybe also `.flowOn(dispatcher)` ?

giannis-rdx commented on app/src/main/java/com/babylon/wallet/android/presentation/account/createaccount/CreateAccountViewModel.kt:
```diff
@@ -227,6 +230,18 @@ class CreateAccountViewModel @Inject constructor(
         )
     }
 
+    private suspend fun shouldCreateWallet(): Boolean {
+        // If profile with networks exists, then no need to create a new one.
+        if (getProfileUseCase.finishedOnboardingProfile() != null) {
+            return false
+        }
+
+        // Delete the instance of the old one, since it might have been an ephemeral profile
+        // created during onboarding while creating an account with ledger, but user killed the app in the meantime.
+        deleteWalletUseCase()
+        return true
+    }
```
💯

giannis-rdx replied:
minor: maybe `shouldCreateNewWallet`?

giannis-rdx commented on app/src/main/java/com/babylon/wallet/android/presentation/onboarding/restore/mnemonics/RestoreMnemonicsViewModel.kt:
```diff
@@ -225,12 +225,18 @@ class RestoreMnemonicsViewModel @Inject constructor(
                             }
                             onRestorationComplete()
                         }.onFailure {
-                            Timber.w(it)
```
maybe to keep the log?

giannis-rdx commented on profile/src/main/java/rdx/works/profile/domain/DeriveProfileUseCase.kt:
```diff
@@ -0,0 +1,51 @@
+        deviceFactorSource: FactorSource.Device,
+        mnemonicWithPassphrase: MnemonicWithPassphrase,
+        accounts: Accounts
+    ): Result<Unit> = when (profileRepository.profileState.first()) {
```
why not the `getProfileUseCase` here?

giannis-rdx reviewed (approved):
Excellent work! 🎩
